graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -438,17 +431,24 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
         within_optional_scope = 'optional' in context and not edge_traversal_is_optional
 
         if edge_traversal_is_optional:
-            # Entering an optional block!
-            # Make sure there's a marked location right before it for the optional Backtrack
-            # to jump back to. Otherwise, the traversal could rewind to an old marked location
-            # and might ignore entire stretches of applied filtering.
-            if context['marked_location_stack'][-1].num_traverses > 0:
+            # Invariant: There must always be a marked location corresponding to the query position
```

Comment:
I see that this invariant is enforced in `ir_sanity_checks.py`, but admittedly it took me awhile to realize that. Up to you but it might be helpful to others to mention that this invariant has a sanity check in place.

Reply:
Good point, I'll mention that in this comment.